core/variant/variant.cpp

```diff
@@ -3254,32 +3254,20 @@ uint32_t Variant::recursive_hash(int recursion_count) const {
 #define hash_compare_scalar(p_lhs, p_rhs) \
 	(hash_compare_scalar_base(p_lhs, p_rhs, true))
 
-#define hash_compare_vector2(p_lhs, p_rhs)        \
-	(hash_compare_scalar((p_lhs).x, (p_rhs).x) && \
-			hash_compare_scalar((p_lhs).y, (p_rhs).y))
-
-#define hash_compare_vector3(p_lhs, p_rhs)               \
-	(hash_compare_scalar((p_lhs).x, (p_rhs).x) &&        \
-			hash_compare_scalar((p_lhs).y, (p_rhs).y) && \
-			hash_compare_scalar((p_lhs).z, (p_rhs).z))
-
-#define hash_compare_vector4(p_lhs, p_rhs)               \
-	(hash_compare_scalar((p_lhs).x, (p_rhs).x) &&        \
-			hash_compare_scalar((p_lhs).y, (p_rhs).y) && \
-			hash_compare_scalar((p_lhs).z, (p_rhs).z) && \
-			hash_compare_scalar((p_lhs).w, (p_rhs).w))
-
-#define hash_compare_quaternion(p_lhs, p_rhs)            \
-	(hash_compare_scalar((p_lhs).x, (p_rhs).x) &&        \
-			hash_compare_scalar((p_lhs).y, (p_rhs).y) && \
-			hash_compare_scalar((p_lhs).z, (p_rhs).z) && \
-			hash_compare_scalar((p_lhs).w, (p_rhs).w))
-
-#define hash_compare_color(p_lhs, p_rhs)                 \
-	(hash_compare_scalar((p_lhs).r, (p_rhs).r) &&        \
-			hash_compare_scalar((p_lhs).g, (p_rhs).g) && \
-			hash_compare_scalar((p_lhs).b, (p_rhs).b) && \
-			hash_compare_scalar((p_lhs).a, (p_rhs).a))
+#define hash_compare_vector2(p_lhs, p_rhs) \
+	(p_lhs).is_semantic_equal(p_rhs)
+
+#define hash_compare_vector3(p_lhs, p_rhs) \
+	(p_lhs).is_semantic_equal(p_rhs)
+
+#define hash_compare_vector4(p_lhs, p_rhs) \
+	(p_lhs).is_semantic_equal(p_rhs)
+
+#define hash_compare_quaternion(p_lhs, p_rhs) \
+	(p_lhs).is_semantic_equal(p_rhs)
+
+#define hash_compare_color(p_lhs, p_rhs) \
+	(p_lhs).is_semantic_equal(p_rhs)
```

Comment:
This is a bit of a weird one, as normally I'd say these defines are redundant & could be removed entirely. However, there's a conditional for if it should use semantic comparison, so it wouldn't be a 1-to-1 replacement. *However*, that conditional isn't actually passed anywhere except the base argument, which only got used on standalone float values (oversight?). The logistics there can be ironed out in another PR though.

Reply:
Only keeping these as they're used by the `Packed*Array` cases, or I'd have used them directly